helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -728,6 +782,100 @@ private void updateDisabledInstances() {
+    List<Long> onlineTimestamps = new ArrayList<>();
+    List<String> historyOnlineList = history.getHistory();
+    if (historyOnlineList != null) {
+      long tailOnlineTime;
+      for (int i = historyOnlineList.size() - 1; i >= 0; i--) {
+        tailOnlineTime =
+            ParticipantHistory.extractTimeFromSessionHistoryString(historyOnlineList.get(i));
+        // Ignore bad format case
+        if (tailOnlineTime != -1) {
+          onlineTimestamps.add(0, tailOnlineTime);
+          // We only need one online timestamp before maintenance mode starts
+          if (tailOnlineTime <= _maintenanceSignal.getTimestamp()) {
+            break;
+          }
+        }
+      }
+    }
+    if (onlineTimestamps.isEmpty() || onlineTimestamps.get(0) > _maintenanceSignal.getTimestamp()) {
+      // Node is a new node in this maintenance mode, no need to time out since no partitions are
+      // assigned to it
+      return false;
+    }
+
+    // Parse offline timestamps from offline list
+    List<Long> offlineTimestamps = new ArrayList<>();
+    List<String> historyOfflineList = history.getOffline();
+    if (historyOfflineList != null) {
+      long tailOfflineTime;
+      for (int i = historyOfflineList.size() - 1; i >= 0; i--) {
+        tailOfflineTime =
+            ParticipantHistory.parseHistoryDateStringToLong(historyOfflineList.get(i));
+        // Ignore bad format case
+        if (tailOfflineTime != -1) {
+          if (tailOfflineTime <= onlineTimestamps.get(0)) {
+            break;
+          }
+          offlineTimestamps.add(0, tailOfflineTime);
+        }
+      }
+    }
+
+    // At this point, onlineTimestamps contains at least 1 timestamp that's before maintenance mode;
+    // offlineTimestamps contains 0+ timestamp that's > the first online timestamp.
+    if (!offlineTimestamps.isEmpty()
+        && offlineTimestamps.get(offlineTimestamps.size() - 1) > onlineTimestamps
+        .get(onlineTimestamps.size() - 1)) {
+      // Make sure for each offline timestamp, 2 online timestamps creates a range around it;
+      // this is usually true unless the timestamps are malformed (missing online timestamps)
+      onlineTimestamps.add(currentTime);
+    }
+
+    // Hop between each pair of online timestamps and find the maximum offline window
+    for (int onlineTimestampIndex = 1, offlineTimestampIndex = 0;
+        onlineTimestampIndex < onlineTimestamps.size() && offlineTimestampIndex < offlineTimestamps
+            .size(); onlineTimestampIndex++) {
+      long onlineTimestamp = onlineTimestamps.get(onlineTimestampIndex);
+      long offlineTimeStamp = offlineTimestamps.get(offlineTimestampIndex);
+
+      // If the offline timestamp isn't within this pair of online timestamp, continue
+      if (offlineTimeStamp > onlineTimestamp) {
+        continue;
+      }
+
+      // Check the largest offline window against the timeout window
+      if (onlineTimestamp - offlineTimeStamp > timeOutWindow) {
```

Comment:
It has the edge case for first interval. It is the interval online time just before maintenance mode, but somehow we dont check whether the offline time is before maintenance.

Let's say if the instance is offline before maintenance we need the potential timeout = online time - Math.max(offlinetimestamp, maintenance entering time);

Reply:
The first offline time is always after the first online time based on my parsing logic, see above. 

Therefore we always start checking by the second online time, and do (second online time - first offline time), for example.

Reply:
Do we really need to go through every pair? Can we just look at the pair with offline time before maintenance entrance and online time after it?

Reply:
@lei-xia We need to handle the case of "offline for too long" during maintenance mode as well, right?